dali/operators/reader/loader/numpy_loader_gpu.cc

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <dirent.h>
+#include <errno.h>
+#include <memory>
+#include <set>
+
+#include "dali/core/common.h"
+#include "dali/operators/reader/loader/numpy_loader_gpu.h"
+
+namespace dali {
+
+// register tensor
+void NumpyLoaderGPU::RegisterTensor(void *buffer, size_t total_size) {
+  if (register_buffers_) {
+    // get raw pointer
+    auto dptr = static_cast<uint8_t*>(buffer);
+    std::unique_lock<std::mutex> reg_lock(reg_mutex_);
+    auto iter = reg_buff_.find(dptr);
+    reg_lock.unlock();
+    if (iter != reg_buff_.end()) {
+      if (iter->second == total_size) {
+        // the right buffer is registered already, return
+        return;
+      } else {
+        cuFileBufDeregister(iter->first);
+      }
+    }
+    // no buffer found or with different size so register again
+    cuFileBufRegister(dptr, total_size, 0);
+    reg_lock.lock();
+    reg_buff_[dptr] = total_size;
+  }
+}
+
+void NumpyLoaderGPU::ReadSampleHelper(CUFileStream *file, ImageFileWrapperGPU& imfile,
+                                      void *buffer, Index offset, size_t total_size) {
+  // register the buffer (if needed)
+  RegisterTensor(buffer, total_size);
```

Comment:
It's not exactly Tensor, it does register the whole tensor list, right? I guess that's why we use the offset into the buffer instead of buffer + offset as a Read target.

Reply:
Renamed